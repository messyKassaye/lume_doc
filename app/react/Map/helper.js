const _mergeArrays = (first, second) => {
  let bigArray;
  let smallArray;
  if (first.length > second.length) {
    bigArray = first;
    smallArray = second;
  } else {
    bigArray = second;
    smallArray = first;
  }

  smallArray.forEach(item => bigArray.push(item));
  return bigArray;
};

const getMarkersBoudingBox = markers =>
  markers.reduce(
    (_b, marker) => {
      const b = _b.slice();
      if (!b[0][0] || marker.longitude < b[0][0]) {
        b[0][0] = marker.longitude;
      }
      if (!b[1][0] || marker.longitude > b[1][0]) {
        b[1][0] = marker.longitude;
      }
      if (!b[1][1] || marker.latitude > b[1][1]) {
        b[1][1] = marker.latitude;
      }
      if (!b[0][1] || marker.latitude < b[0][1]) {
        b[0][1] = marker.latitude;
      }
      return b;
    },
    [
      [null, null],
      [null, null],
    ]
  );

const markersToStyleFormat = markers =>
  markers.map((marker, index) => {
    const properties = marker.properties || {};
    const { longitude, latitude } = marker;
    properties.index = index;
    return {
      type: 'Feature',
      properties,
      geometry: {
        type: 'Point',
        coordinates: [longitude, latitude],
      },
    };
  });

const noop = () => {};

const extractPropertyBasedMarkers = (entity, geolocationProp, propertyMetadata, color) => {
  const markers = [];
  propertyMetadata
    .filter(mo => mo.get('value'))
    .forEach(point => {
      markers.push({
        properties: { entity: entity.toJS(), color, info: point.getIn(['value', 'label']) },
        latitude: point.getIn(['value', 'lat']),
        longitude: point.getIn(['value', 'lon']),
        label: geolocationProp.get('label'),
      });
    });
  return markers;
};

const extractInheritedMarkers = (entity, geolocationProp, propertyMetadata, color) => {
  const markers = [];
  propertyMetadata.forEach(relatedProperty => {
    if (relatedProperty.has('inherit_geolocation')) {
      // conditional is a quick hack to prevent app crash
      relatedProperty
        .get('inherit_geolocation')
        .filter(mo => mo.get('value'))
        .forEach(point => {
          markers.push({
            properties: {
              entity: entity.toJS(),
              color,
              info: point.getIn(['value', 'label']),
              inherited: true,
              label: relatedProperty.get('label'),
              context: geolocationProp.get('content'),
            },
            latitude: point.getIn(['value', 'lat']),
            longitude: point.getIn(['value', 'lon']),
            label: geolocationProp.get('label'),
          });
        });
    }
  });
  return markers;
};

const extractMarkers = ({ entity, geolocationProps, metadata, color, type }) => {
  let markers = [];
  const geolocationPropNames = geolocationProps.map(prop => prop.get('name'));

  if (geolocationProps.size && metadata) {
    const [...properties] = metadata.keys();
    properties.forEach(property => {
      if (geolocationPropNames.includes(property) && metadata.has(property)) {
        const geolocationProp = geolocationProps.find(p => p.get('name') === property);
        const propertyMetadata = metadata.get(property);
        if (type === 'PropertyBased') {
          markers = _mergeArrays(
            markers,
            extractPropertyBasedMarkers(entity, geolocationProp, propertyMetadata, color)
          );
        } else {
          markers = _mergeArrays(
            markers,
            extractInheritedMarkers(entity, geolocationProp, propertyMetadata, color)
          );
        }
      }
    });
  }
  return markers;
};

const getPropertyBasedMarkers = (template, entity) => {
  const color = template.get('color');
  const templateProperties = template.get('properties');
  const metadata = entity.get('metadata');

  const geolocationProps = templateProperties.filter(
    property => property.get('type') === 'geolocation'
  );

  return extractMarkers({ entity, geolocationProps, metadata, color, type: 'PropertyBased' });
};

const getInheritedMarkers = (template, entity, templates) => {
  const templateProperties = template.get('properties');
  const color = template.get('color');
  const metadata = entity.get('metadata');

  const geolocationProps = templateProperties.filter(property => {
    if (property.get('type') === 'relationship' && property.has('inheritProperty')) {
      const contentTemplate = templates.find(
        t => t.get('_id').toString() === property.get('content').toString()
      );
      const inheritedProperty = contentTemplate
        .get('properties')
        .find(p => p.get('_id').toString() === property.get('inheritProperty').toString());
      return inheritedProperty.get('type') === 'geolocation';
    }
    return false;
  });

  return extractMarkers({ entity, geolocationProps, metadata, color, type: 'Inherited' });
};

const getEntityMarkers = (entity, templates) => {
  const template = templates.find(_t => _t.get('_id') === entity.get('template'));

  const propertyBasedMarkers = getPropertyBasedMarkers(template, entity);
  const inheritedMarkers = getInheritedMarkers(template, entity, templates);

  return _mergeArrays(propertyBasedMarkers, inheritedMarkers);
};

const getMarkers = (entities, templates) => {
  const m = entities.reduce((markers, entity) => {
    const entityMarkers = getEntityMarkers(entity, templates);
    return _mergeArrays(markers, entityMarkers);
  }, []);
  return m;
};

const TRANSITION_PROPS = {
  transitionDuration: 0,
  transitionEasing: t => t,
  transitionInterruption: 1,
  onTransitionStart: noop,
  onTransitionInterrupt: noop,
  onTransitionEnd: noop,
  onViewportChange: noop,
  onStateChange: noop,
};

export { getMarkersBoudingBox, markersToStyleFormat, getMarkers, TRANSITION_PROPS };
