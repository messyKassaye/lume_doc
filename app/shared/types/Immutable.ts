import Immutable from 'immutable';

export type IImmutable<T> = T extends string
  ? string
  : T extends number
  ? number
  : T extends boolean
  ? boolean
  : T extends null
  ? null
  : T extends undefined
  ? undefined
  : T extends Array<infer Elem>
  ? Immutable.List<IImmutable<Elem>>
  : {
      toJS(): T;
      get<Field extends keyof T>(_field: Field): IImmutable<T[Field]>;
    };
