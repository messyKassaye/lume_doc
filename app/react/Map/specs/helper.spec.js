/** @format */

import Immutable from 'immutable';

import { getMarkersBoudingBox, markersToStyleFormat, getMarkers } from '../helper';

describe('map helper', () => {
  const markers = [
    { latitude: 2, longitude: 32 },
    { latitude: 23, longitude: 21 },
    { latitude: 7, longitude: 11 },
    { latitude: 22, longitude: -21 },
  ];

  describe('getMarkersBoudingBox()', () => {
    it('should parse all the markers and return the bounding box for them', () => {
      expect(getMarkersBoudingBox(markers)).toEqual([
        [-21, 2],
        [32, 23],
      ]);
    });
  });

  describe('markersToStyleFormat()', () => {
    it('should parse all the markers and return the bounding box for them', () => {
      const expectedFormat = [
        {
          geometry: { coordinates: [32, 2], type: 'Point' },
          properties: { index: 0 },
          type: 'Feature',
        },
        {
          geometry: { coordinates: [21, 23], type: 'Point' },
          properties: { index: 1 },
          type: 'Feature',
        },
        {
          geometry: { coordinates: [11, 7], type: 'Point' },
          properties: { index: 2 },
          type: 'Feature',
        },
        {
          geometry: { coordinates: [-21, 22], type: 'Point' },
          properties: { index: 3 },
          type: 'Feature',
        },
      ];
      expect(markersToStyleFormat(markers)).toEqual(expectedFormat);
    });
  });

  describe('getMarkers()', () => {
    let templates;
    let entities;

    beforeEach(() => {
      templates = Immutable.fromJS([
        {
          _id: 't1',
          properties: [
            { _id: 't1p1', type: 'geolocation', name: 'geoProperty', label: 'geoPropertyLabel' },
            { type: 'geolocation', name: 'secondGeoProperty', label: 'secondGeoPropertyLabel' },
            {
              type: 'relationship',
              name: 'inheritedGeo1',
              label: 'Inherited Property Label 1',
              content: 't2',
              inheritProperty: 't2p1',
            },
            {
              type: 'relationship',
              name: 'inheritedGeo2',
              label: 'Inherited Property Label 2',
              content: 't4',
              inheritProperty: 't4p2',
            },
          ],
          color: 'red',
        },
        {
          _id: 't2',
          properties: [
            {
              _id: 't2p1',
              type: 'geolocation',
              name: 'anotherGeoProperty',
              label: 'anotherGeoPropertyLabel',
            },
          ],
          color: 'blue',
        },
        {
          _id: 't3',
          properties: [{ type: 'notGeolocation', name: 'notGeo' }],
          color: 'green',
        },
        {
          _id: 't4',
          properties: [
            {
              _id: 't4p1',
              type: 'relationship',
              name: 'anotherInheritedGeo',
              label: 'Another Inherited Property Label',
              content: 't1',
              inheritProperty: 't1p1',
            },
            {
              _id: 't4p2',
              type: 'geolocation',
              name: 't4GeoProperty',
              label: 't4GeoPropertyLabel',
            },
          ],
          color: 'yellow',
        },
      ]);

      entities = Immutable.fromJS([
        {
          sharedId: 'e1',
          template: 't1',
          metadata: {
            geoProperty: [
              { value: { lat: 7, lon: 13 } },
              { value: { lat: 13, lon: 7, label: 'home' } },
            ],
            secondGeoProperty: [{ value: { lat: 5, lon: 13 } }],
            text: [{ value: 'text value' }],
          },
        },
        {
          template: 't1',
          metadata: {
            geoProperty: [{ value: { lat: 5, lon: 22 } }],
            secondGeoProperty: [{ value: null }],
          },
        },
        { template: 't3', metadata: { notGeo: [{ value: { lat: 1977, lon: 7 } }] } },
        {
          template: 't2',
          sharedId: 'e4',
          title: 'e4Title',
          metadata: { anotherGeoProperty: [{ value: { lat: 2018, lon: 6 } }] },
        },
        { template: 't3' },
        { template: 't2' },
      ]);
    });

    it('should calculate the entity markers based on template data', () => {
      const processedMarkers = getMarkers(entities, templates);
      expect(processedMarkers.length).toBe(5);
      expect(processedMarkers).toMatchSnapshot();
    });

    it('should return an empty array if no data', () => {
      const processedMarkers = getMarkers(Immutable.fromJS([]), templates);
      expect(processedMarkers).toEqual([]);
    });

    describe('Inheritance', () => {
      beforeEach(() => {
        entities = entities
          .setIn(
            [0, 'metadata', 'inheritedGeo1'],
            Immutable.fromJS([
              {
                value: 'e5',
                label: 'e5 Label',
                inherit_geolocation: [{ value: { lat: 23, lon: 1946 } }],
              },
              {
                value: 'e4',
                label: 'e4 Label',
                inherit_geolocation: [
                  { value: { lat: 2018, lon: 6 } },
                  { value: { lat: 1977, lon: 1982, label: 'inherited label' } },
                ],
              },
            ])
          )
          .setIn(
            [0, 'metadata', 'inheritedGeo2'],
            Immutable.fromJS([
              {
                value: 'e5',
                label: 'e5 Label',
                inherit_geolocation: [{ value: { lat: 23, lon: 1946 } }],
              },
            ])
          )
          .push(
            Immutable.fromJS({
              template: 't2',
              sharedId: 'e5',
              metadata: { anotherGeoProperty: [{ value: { lat: 23, lon: 1946 } }] },
            })
          )
          .push(
            Immutable.fromJS({
              template: 't4',
              metadata: {
                anotherInheritedGeo: [
                  {
                    value: 'e1',
                    label: 'e1 Label',
                    inherit_geolocation: [{ value: { lat: 13, lon: 7, label: 'home' } }],
                  },
                ],
              },
            })
          );
      });

      it('should include inherited properties', () => {
        const processedMarkers = getMarkers(entities, templates);
        expect(processedMarkers.length).toBe(11);
        expect(processedMarkers).toEqual(
          expect.arrayContaining([
            expect.objectContaining({ properties: expect.objectContaining({ inherited: true }) }),
          ])
        );
        expect(processedMarkers).toMatchSnapshot();
      });
    });
  });
});
