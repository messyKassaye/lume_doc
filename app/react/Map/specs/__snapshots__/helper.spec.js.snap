// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map helper getMarkers() Inheritance should include inherited properties 1`] = `
Array [
  Object {
    "label": "Inherited Property Label 1",
    "latitude": 23,
    "longitude": 1946,
    "properties": Object {
      "color": "red",
      "context": "t2",
      "entity": Object {
        "metadata": Object {
          "geoProperty": Array [
            Object {
              "value": Object {
                "lat": 7,
                "lon": 13,
              },
            },
            Object {
              "value": Object {
                "label": "home",
                "lat": 13,
                "lon": 7,
              },
            },
          ],
          "inheritedGeo1": Array [
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 23,
                    "lon": 1946,
                  },
                },
              ],
              "label": "e5 Label",
              "value": "e5",
            },
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 2018,
                    "lon": 6,
                  },
                },
                Object {
                  "value": Object {
                    "label": "inherited label",
                    "lat": 1977,
                    "lon": 1982,
                  },
                },
              ],
              "label": "e4 Label",
              "value": "e4",
            },
          ],
          "inheritedGeo2": Array [
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 23,
                    "lon": 1946,
                  },
                },
              ],
              "label": "e5 Label",
              "value": "e5",
            },
          ],
          "secondGeoProperty": Array [
            Object {
              "value": Object {
                "lat": 5,
                "lon": 13,
              },
            },
          ],
          "text": Array [
            Object {
              "value": "text value",
            },
          ],
        },
        "sharedId": "e1",
        "template": "t1",
      },
      "info": undefined,
      "inherited": true,
      "label": "e5 Label",
    },
  },
  Object {
    "label": "Inherited Property Label 1",
    "latitude": 2018,
    "longitude": 6,
    "properties": Object {
      "color": "red",
      "context": "t2",
      "entity": Object {
        "metadata": Object {
          "geoProperty": Array [
            Object {
              "value": Object {
                "lat": 7,
                "lon": 13,
              },
            },
            Object {
              "value": Object {
                "label": "home",
                "lat": 13,
                "lon": 7,
              },
            },
          ],
          "inheritedGeo1": Array [
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 23,
                    "lon": 1946,
                  },
                },
              ],
              "label": "e5 Label",
              "value": "e5",
            },
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 2018,
                    "lon": 6,
                  },
                },
                Object {
                  "value": Object {
                    "label": "inherited label",
                    "lat": 1977,
                    "lon": 1982,
                  },
                },
              ],
              "label": "e4 Label",
              "value": "e4",
            },
          ],
          "inheritedGeo2": Array [
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 23,
                    "lon": 1946,
                  },
                },
              ],
              "label": "e5 Label",
              "value": "e5",
            },
          ],
          "secondGeoProperty": Array [
            Object {
              "value": Object {
                "lat": 5,
                "lon": 13,
              },
            },
          ],
          "text": Array [
            Object {
              "value": "text value",
            },
          ],
        },
        "sharedId": "e1",
        "template": "t1",
      },
      "info": undefined,
      "inherited": true,
      "label": "e4 Label",
    },
  },
  Object {
    "label": "Inherited Property Label 1",
    "latitude": 1977,
    "longitude": 1982,
    "properties": Object {
      "color": "red",
      "context": "t2",
      "entity": Object {
        "metadata": Object {
          "geoProperty": Array [
            Object {
              "value": Object {
                "lat": 7,
                "lon": 13,
              },
            },
            Object {
              "value": Object {
                "label": "home",
                "lat": 13,
                "lon": 7,
              },
            },
          ],
          "inheritedGeo1": Array [
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 23,
                    "lon": 1946,
                  },
                },
              ],
              "label": "e5 Label",
              "value": "e5",
            },
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 2018,
                    "lon": 6,
                  },
                },
                Object {
                  "value": Object {
                    "label": "inherited label",
                    "lat": 1977,
                    "lon": 1982,
                  },
                },
              ],
              "label": "e4 Label",
              "value": "e4",
            },
          ],
          "inheritedGeo2": Array [
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 23,
                    "lon": 1946,
                  },
                },
              ],
              "label": "e5 Label",
              "value": "e5",
            },
          ],
          "secondGeoProperty": Array [
            Object {
              "value": Object {
                "lat": 5,
                "lon": 13,
              },
            },
          ],
          "text": Array [
            Object {
              "value": "text value",
            },
          ],
        },
        "sharedId": "e1",
        "template": "t1",
      },
      "info": "inherited label",
      "inherited": true,
      "label": "e4 Label",
    },
  },
  Object {
    "label": "Inherited Property Label 2",
    "latitude": 23,
    "longitude": 1946,
    "properties": Object {
      "color": "red",
      "context": "t4",
      "entity": Object {
        "metadata": Object {
          "geoProperty": Array [
            Object {
              "value": Object {
                "lat": 7,
                "lon": 13,
              },
            },
            Object {
              "value": Object {
                "label": "home",
                "lat": 13,
                "lon": 7,
              },
            },
          ],
          "inheritedGeo1": Array [
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 23,
                    "lon": 1946,
                  },
                },
              ],
              "label": "e5 Label",
              "value": "e5",
            },
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 2018,
                    "lon": 6,
                  },
                },
                Object {
                  "value": Object {
                    "label": "inherited label",
                    "lat": 1977,
                    "lon": 1982,
                  },
                },
              ],
              "label": "e4 Label",
              "value": "e4",
            },
          ],
          "inheritedGeo2": Array [
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 23,
                    "lon": 1946,
                  },
                },
              ],
              "label": "e5 Label",
              "value": "e5",
            },
          ],
          "secondGeoProperty": Array [
            Object {
              "value": Object {
                "lat": 5,
                "lon": 13,
              },
            },
          ],
          "text": Array [
            Object {
              "value": "text value",
            },
          ],
        },
        "sharedId": "e1",
        "template": "t1",
      },
      "info": undefined,
      "inherited": true,
      "label": "e5 Label",
    },
  },
  Object {
    "label": "geoPropertyLabel",
    "latitude": 7,
    "longitude": 13,
    "properties": Object {
      "color": "red",
      "entity": Object {
        "metadata": Object {
          "geoProperty": Array [
            Object {
              "value": Object {
                "lat": 7,
                "lon": 13,
              },
            },
            Object {
              "value": Object {
                "label": "home",
                "lat": 13,
                "lon": 7,
              },
            },
          ],
          "inheritedGeo1": Array [
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 23,
                    "lon": 1946,
                  },
                },
              ],
              "label": "e5 Label",
              "value": "e5",
            },
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 2018,
                    "lon": 6,
                  },
                },
                Object {
                  "value": Object {
                    "label": "inherited label",
                    "lat": 1977,
                    "lon": 1982,
                  },
                },
              ],
              "label": "e4 Label",
              "value": "e4",
            },
          ],
          "inheritedGeo2": Array [
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 23,
                    "lon": 1946,
                  },
                },
              ],
              "label": "e5 Label",
              "value": "e5",
            },
          ],
          "secondGeoProperty": Array [
            Object {
              "value": Object {
                "lat": 5,
                "lon": 13,
              },
            },
          ],
          "text": Array [
            Object {
              "value": "text value",
            },
          ],
        },
        "sharedId": "e1",
        "template": "t1",
      },
      "info": undefined,
    },
  },
  Object {
    "label": "geoPropertyLabel",
    "latitude": 13,
    "longitude": 7,
    "properties": Object {
      "color": "red",
      "entity": Object {
        "metadata": Object {
          "geoProperty": Array [
            Object {
              "value": Object {
                "lat": 7,
                "lon": 13,
              },
            },
            Object {
              "value": Object {
                "label": "home",
                "lat": 13,
                "lon": 7,
              },
            },
          ],
          "inheritedGeo1": Array [
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 23,
                    "lon": 1946,
                  },
                },
              ],
              "label": "e5 Label",
              "value": "e5",
            },
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 2018,
                    "lon": 6,
                  },
                },
                Object {
                  "value": Object {
                    "label": "inherited label",
                    "lat": 1977,
                    "lon": 1982,
                  },
                },
              ],
              "label": "e4 Label",
              "value": "e4",
            },
          ],
          "inheritedGeo2": Array [
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 23,
                    "lon": 1946,
                  },
                },
              ],
              "label": "e5 Label",
              "value": "e5",
            },
          ],
          "secondGeoProperty": Array [
            Object {
              "value": Object {
                "lat": 5,
                "lon": 13,
              },
            },
          ],
          "text": Array [
            Object {
              "value": "text value",
            },
          ],
        },
        "sharedId": "e1",
        "template": "t1",
      },
      "info": "home",
    },
  },
  Object {
    "label": "secondGeoPropertyLabel",
    "latitude": 5,
    "longitude": 13,
    "properties": Object {
      "color": "red",
      "entity": Object {
        "metadata": Object {
          "geoProperty": Array [
            Object {
              "value": Object {
                "lat": 7,
                "lon": 13,
              },
            },
            Object {
              "value": Object {
                "label": "home",
                "lat": 13,
                "lon": 7,
              },
            },
          ],
          "inheritedGeo1": Array [
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 23,
                    "lon": 1946,
                  },
                },
              ],
              "label": "e5 Label",
              "value": "e5",
            },
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 2018,
                    "lon": 6,
                  },
                },
                Object {
                  "value": Object {
                    "label": "inherited label",
                    "lat": 1977,
                    "lon": 1982,
                  },
                },
              ],
              "label": "e4 Label",
              "value": "e4",
            },
          ],
          "inheritedGeo2": Array [
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "lat": 23,
                    "lon": 1946,
                  },
                },
              ],
              "label": "e5 Label",
              "value": "e5",
            },
          ],
          "secondGeoProperty": Array [
            Object {
              "value": Object {
                "lat": 5,
                "lon": 13,
              },
            },
          ],
          "text": Array [
            Object {
              "value": "text value",
            },
          ],
        },
        "sharedId": "e1",
        "template": "t1",
      },
      "info": undefined,
    },
  },
  Object {
    "label": "geoPropertyLabel",
    "latitude": 5,
    "longitude": 22,
    "properties": Object {
      "color": "red",
      "entity": Object {
        "metadata": Object {
          "geoProperty": Array [
            Object {
              "value": Object {
                "lat": 5,
                "lon": 22,
              },
            },
          ],
          "secondGeoProperty": Array [
            Object {
              "value": null,
            },
          ],
        },
        "template": "t1",
      },
      "info": undefined,
    },
  },
  Object {
    "label": "anotherGeoPropertyLabel",
    "latitude": 2018,
    "longitude": 6,
    "properties": Object {
      "color": "blue",
      "entity": Object {
        "metadata": Object {
          "anotherGeoProperty": Array [
            Object {
              "value": Object {
                "lat": 2018,
                "lon": 6,
              },
            },
          ],
        },
        "sharedId": "e4",
        "template": "t2",
        "title": "e4Title",
      },
      "info": undefined,
    },
  },
  Object {
    "label": "anotherGeoPropertyLabel",
    "latitude": 23,
    "longitude": 1946,
    "properties": Object {
      "color": "blue",
      "entity": Object {
        "metadata": Object {
          "anotherGeoProperty": Array [
            Object {
              "value": Object {
                "lat": 23,
                "lon": 1946,
              },
            },
          ],
        },
        "sharedId": "e5",
        "template": "t2",
      },
      "info": undefined,
    },
  },
  Object {
    "label": "Another Inherited Property Label",
    "latitude": 13,
    "longitude": 7,
    "properties": Object {
      "color": "yellow",
      "context": "t1",
      "entity": Object {
        "metadata": Object {
          "anotherInheritedGeo": Array [
            Object {
              "inherit_geolocation": Array [
                Object {
                  "value": Object {
                    "label": "home",
                    "lat": 13,
                    "lon": 7,
                  },
                },
              ],
              "label": "e1 Label",
              "value": "e1",
            },
          ],
        },
        "template": "t4",
      },
      "info": "home",
      "inherited": true,
      "label": "e1 Label",
    },
  },
]
`;

exports[`map helper getMarkers() should calculate the entity markers based on template data 1`] = `
Array [
  Object {
    "label": "geoPropertyLabel",
    "latitude": 7,
    "longitude": 13,
    "properties": Object {
      "color": "red",
      "entity": Object {
        "metadata": Object {
          "geoProperty": Array [
            Object {
              "value": Object {
                "lat": 7,
                "lon": 13,
              },
            },
            Object {
              "value": Object {
                "label": "home",
                "lat": 13,
                "lon": 7,
              },
            },
          ],
          "secondGeoProperty": Array [
            Object {
              "value": Object {
                "lat": 5,
                "lon": 13,
              },
            },
          ],
          "text": Array [
            Object {
              "value": "text value",
            },
          ],
        },
        "sharedId": "e1",
        "template": "t1",
      },
      "info": undefined,
    },
  },
  Object {
    "label": "geoPropertyLabel",
    "latitude": 13,
    "longitude": 7,
    "properties": Object {
      "color": "red",
      "entity": Object {
        "metadata": Object {
          "geoProperty": Array [
            Object {
              "value": Object {
                "lat": 7,
                "lon": 13,
              },
            },
            Object {
              "value": Object {
                "label": "home",
                "lat": 13,
                "lon": 7,
              },
            },
          ],
          "secondGeoProperty": Array [
            Object {
              "value": Object {
                "lat": 5,
                "lon": 13,
              },
            },
          ],
          "text": Array [
            Object {
              "value": "text value",
            },
          ],
        },
        "sharedId": "e1",
        "template": "t1",
      },
      "info": "home",
    },
  },
  Object {
    "label": "secondGeoPropertyLabel",
    "latitude": 5,
    "longitude": 13,
    "properties": Object {
      "color": "red",
      "entity": Object {
        "metadata": Object {
          "geoProperty": Array [
            Object {
              "value": Object {
                "lat": 7,
                "lon": 13,
              },
            },
            Object {
              "value": Object {
                "label": "home",
                "lat": 13,
                "lon": 7,
              },
            },
          ],
          "secondGeoProperty": Array [
            Object {
              "value": Object {
                "lat": 5,
                "lon": 13,
              },
            },
          ],
          "text": Array [
            Object {
              "value": "text value",
            },
          ],
        },
        "sharedId": "e1",
        "template": "t1",
      },
      "info": undefined,
    },
  },
  Object {
    "label": "geoPropertyLabel",
    "latitude": 5,
    "longitude": 22,
    "properties": Object {
      "color": "red",
      "entity": Object {
        "metadata": Object {
          "geoProperty": Array [
            Object {
              "value": Object {
                "lat": 5,
                "lon": 22,
              },
            },
          ],
          "secondGeoProperty": Array [
            Object {
              "value": null,
            },
          ],
        },
        "template": "t1",
      },
      "info": undefined,
    },
  },
  Object {
    "label": "anotherGeoPropertyLabel",
    "latitude": 2018,
    "longitude": 6,
    "properties": Object {
      "color": "blue",
      "entity": Object {
        "metadata": Object {
          "anotherGeoProperty": Array [
            Object {
              "value": Object {
                "lat": 2018,
                "lon": 6,
              },
            },
          ],
        },
        "sharedId": "e4",
        "template": "t2",
        "title": "e4Title",
      },
      "info": undefined,
    },
  },
]
`;
